_posts/2022-08-17-using-the-kani-rust-verifier-on-tokio-bytes.md

```diff
@@ -0,0 +1,567 @@
+    - [A proof about `split_off`](#a-proof-about-split_off)
+    - [About specifications](#about-specifications)
+  - [Summary](#summary)
+
+## Diving into `Bytes::BytesMut`
+
+At the bottom of the Tokio stack is the [Bytes](https://github.com/tokio-rs/bytes) library, a _"rich set of utilities for manipulating byte arrays"_, which reflects the fact that networking applications, at their core, have to manipulate byte streams.
+Inside the library are two container types:
+- `Bytes`: a cheaply cloneable and sliceable chunk of contiguous memory
+- `BytesMut`: a unique reference to a (potentially shared) contiguous slice of memory
+  
+As the names suggest, the bytes pointed to by a `Bytes` instance are immutable whereas they are mutable through a `BytesMut` instance.
+In this post we'll focus on `BytesMut`.
+Let's start with its definition:
+
+```rust
+pub struct BytesMut {
+    ptr: NonNull<u8>,
+    len: usize,
+    cap: usize,
+    data: *mut Shared,
+}
+
+struct Shared {
+    vec: Vec<u8>,
+    original_capacity_repr: usize,
+    ref_count: AtomicUsize,
+}
+```
+
+The straightforward part is the first three fields.
+Similar to a `Vec<u8>`, we have a pointer `ptr` to the backing buffer of size `cap` bytes with `len` bytes currently in use.
+The field `data` is a bit trickier because it handles sharing of the backing buffer.
+In particular, `data` is used to distinguish between two kinds of representation: `KIND_VEC` and `KIND_ARC`.
+- A `KIND_VEC` is used when the backing buffer is only pointed to by `ptr`.
+In this case, `data` is not used as a pointer but instead contains a sentinel value (the bottom bit is set to `'0b1`) to mark that we're using this representation.
+- A `KIND_ARC` is used when the backing buffer is shared by multiple `BytesMut` instances.
+In this case, each instance's `data` field points to a single `Shared` object whose `ref_count` keeps track of the number of aliases.
+In order for this aliasing to be safe, the implementation must ensure that each instance `ptr` field points to a disjoint part of the slice.
+[Additionally, the rules for [pointer alignment](https://doc.rust-lang.org/reference/type-layout.html#pointers-and-references-layout) ensure that we will never mix up representations (since the bottom bit of a pointer will always be `0`).]
+
+As a user of `BytesMut` you don't need to worry about this distinction.
+Under the hood, the implementation will switch representations as required [and, within the library implementation, we can use the private method `kind()` to query the representation of a given instance].
+A simplified state machine of when the representation changes is as follows:
+
+```
+  │
+  │ new()
+  ▼
+┌──────────┐  split_off()   ┌──────────┐
+│ KIND_VEC │ ─────────────► │ KIND_ARC │
+└──────────┘                └──────────┘
+  ▲                           │
+  │                           │
+  └───────────────────────────┘
+                    reserve()
+```
```

Comment:
I'm happy with this diagram, but I am wondering if it might be better to draw it using something other than unicode chars.

Reply:
Thanks, added an image to replace this diagram. Please take a look!

Reply:
LGTM. If you already have an assets directory, then I think it is fine to put it up there.

For markdown only files, I think something like https://github.blog/2022-02-14-include-diagrams-markdown-files-mermaid/ can be used to avoid pushing images to the repo.